### PR TITLE
Changing the mock server to use port reserver insted of hard ports

### PR DESCRIPTION
**Pull request by MeniZalzman**
Changing the Server mock to used port reserver instead of hard coding the port in the tests, This change also enable executing this tests in parallel as the port reserver making sure not to allocate reserved ports to other tests.

nuget/home#1050

@deepakaravindr @yishaigalatzer @emgarten @feiling

### CODE REVIEW

**Comment by yishaigalatzer** [comment]
Does this change also include fixing all the consistently broken test you mentioned today?

@yishaigalatzer Yes all the tests are passing

File: test/NuGet.CommandLine.Test/CommandRunner.cs [hunk 1]
```diff
@@ -12,7 +12,7 @@ public class CommandRunner
     {
         // Item1 of the returned tuple is the exit code. Item2 is the standard output, and Item3 
         // is the error output.
-        public static Tuple<int, string, string> Run(string process, string workingDirectory, string arguments, bool waitForExit, int timeOutInMilliseconds = 60000,
+        public static Tuple<int, string, string> Run(string process, string workingDirectory, string arguments, bool waitForExit, int timeOutInMilliseconds = 600000000,
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/CommandRunner.cs]: 1. Please while you change code, make the lines shorter (no longer than 120 chars). It makes reviewing a PR much easier. The easy way to see this is in a github PR, you never want to scroll.
2. Why is the timeout changing? Unless there is a very good reason it shouldn't because a failed test will halt the build for a long time

File: test/NuGet.CommandLine.Test/MockServer.cs [hunk 2]
```diff
@@ -15,28 +15,29 @@ namespace NuGet.CommandLine.Test
...
 
         /// <summary>
         /// Initializes an instance of MockServer.
         /// </summary>
-        /// <param name="endPoint">The endpoint of the server.</param>
-        public MockServer(string endPoint)
+        public MockServer()
         {
-            _endPoint = endPoint;
+            this._portReserver = new PortReserver();
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/MockServer.cs]: don't use `this.`

File: test/NuGet.CommandLine.Test/MockServer.cs [hunk 3]
```diff
@@ -286,7 +289,7 @@ void HandleRequest()
         public string ToODataFeed(IEnumerable<IPackage> packages, string title)
         {
             string nsAtom = "http://www.w3.org/2005/Atom";
-            var id = string.Format(CultureInfo.InvariantCulture, "{0}{1}", _endPoint, title);
+            var id = string.Format(CultureInfo.InvariantCulture, "{0}{1}", this.Uri, title);
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/MockServer.cs]: this (and other places as well)

File: test/NuGet.CommandLine.Test/MockServer.cs [hunk 4]
```diff
@@ -340,6 +343,28 @@ public string ToOData(IPackage package)
...
+
+                    // Disposing the PortReserver
+                    this._portReserver.Dispose();
+                }
+                
+                _disposedValue = true;
+            }
+        }
+        
+        public void Dispose()
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/MockServer.cs:364]: this pattern is unnecessary here, just use dispose.

File: test/NuGet.CommandLine.Test/NuGetDeleteCommandTest.cs [hunk 5]
```diff
@@ -85,33 +85,33 @@ public void DeleteCommand_DeleteFromHttpSource()
...
                     "delete", "testPackage1", "1.1.0",
-                    "-Source", mockServerEndPoint + "nuget", "-NonInteractive" };
-            var r = CommandRunner.Run(
-                nugetexe,
-                Directory.GetCurrentDirectory(),
-                string.Join(" ", args),
-                waitForExit: true);
-            server.Stop();
+                    "-Source", server.Uri + "nuget", "-NonInteractive" };
+                var r = CommandRunner.Run(
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/NuGetDeleteCommandTest.cs:105]: is there a race here?

File: test/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs [hunk 6]
```diff
@@ -456,23 +456,25 @@ public void InstallCommand_GetLastestPrereleaseVersion()
...
+                    server.Start();
+
+                    // Act
+                    var args = "install testPackage1 -Prerelease -Source " + server.Uri + "nuget";
+                    var r1 = CommandRunner.Run(
+                        nugetexe,
+                        workingDirectory,
+                        args,
+                        waitForExit: true);
+                    server.Stop();
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: new line above `server.stop` it looks like part of the Run this way

File: test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 7]
```diff
@@ -392,70 +396,72 @@ public void PushCommand_PushToServerBasicAuth()
...
-
-                    if (credential.Equals("testuser:testpassword", StringComparison.OrdinalIgnoreCase))
+                    server.Listener.AuthenticationSchemes = AuthenticationSchemes.Basic;
+                    server.Get.Add("/nuget", r =>
                     {
-                        res.StatusCode = (int)HttpStatusCode.OK;
-                    }
-                    else
+                        var h = r.Headers["Authorization"];
+                        var credential = System.Text.Encoding.Default.GetString(Convert.FromBase64String(h.Substring(6)));
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs]: drop unnecessary namespace and move to a using statement if possible

File: test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 8]
```diff
@@ -392,70 +396,72 @@ public void PushCommand_PushToServerBasicAuth()
...
+                        else
+                        {
+                            res.AddHeader("WWW-Authenticate", "Basic ");
+                            res.StatusCode = (int)HttpStatusCode.Unauthorized;
+                        }
+                    }));
+                    server.Start();
+
+                    // Act
+                    Environment.SetEnvironmentVariable("FORCE_NUGET_EXE_INTERACTIVE", "true");
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:428]: this looks like a problem waiting to happen. There is a race between tests here. This should be protected by a mutex.

File: test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 9]
```diff
@@ -483,70 +488,72 @@ public void PushCommand_PushToServerBasicAuthDisableBuffering()
...
```
Comment (yishaigalatzer) [inline test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:520]: same comment as above about potential race. Also requires proper cleanup

**Comment by yishaigalatzer** [comment]
:watch:

File: test/NuGet.CommandLine.Test/CommandRunner.cs (hunk 1 above)
Comment (MeniZalzman) [inline test/NuGet.CommandLine.Test/CommandRunner.cs]: good catch, I change the timeout for debugging purposes, will revert this change

File: test/NuGet.CommandLine.Test/MockServer.cs (hunk 2 above)
Comment (MeniZalzman) [inline test/NuGet.CommandLine.Test/MockServer.cs]: Removed

File: test/NuGet.CommandLine.Test/MockServer.cs (hunk 3 above)
Comment (MeniZalzman) [inline test/NuGet.CommandLine.Test/MockServer.cs]: Removed

File: test/NuGet.CommandLine.Test/MockServer.cs (hunk 4 above)
Comment (MeniZalzman) [inline test/NuGet.CommandLine.Test/MockServer.cs:364]: I prefer to use this dispose pattern

File: test/NuGet.CommandLine.Test/NuGetDeleteCommandTest.cs (hunk 5 above)
Comment (MeniZalzman) [inline test/NuGet.CommandLine.Test/NuGetDeleteCommandTest.cs:105]: not that I see, where do you think we are expose to race condition?

File: test/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs (hunk 6 above)
Comment (MeniZalzman) [inline test/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: Right because of that I missed it, no need to call server stop anymore it's part of the dispose method

File: test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs (hunk 7 above)
Comment (MeniZalzman) [inline test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs]: done

File: test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs (hunk 8 above)
Comment (MeniZalzman) [inline test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:428]: @yishaigalatzer  Can we split this change to a different code review?

**Comment by deepakaravindr** [comment]
Looks good to me
